### PR TITLE
[TMVA] Fix RBDT evaluation of XGBoost models by merging RBDT with FastForest

### PR DESCRIPTION
**Pull request by guitargeek**
Consolidate RBDT as specified in the ROOT plan of work 2024.

The backends of RBDT are replaced with a single new backend:
the logic from the FastForest library:
https://github.com/guitargeek/XGBoost-FastForest

The logic in that library was originally taken from the GBRForest in
CMSSW:
https://github.com/cms-sw/cmssw/blob/master/CommonTools/MVAUtils/interface/GBRForestTools.h

The interface remains the same, only that the template parameter
specifying the backend is gone.

This change adds support for unbalanced trees.

To see the fixed logic, as one run a modified version of the `tmva101_Training.py` tutorial (remember to run `tutorials/tmva/tmva100_DataPreparation.py` first):
```python
import ROOT
import numpy as np

from tmva100_DataPreparation import variables

def load_data(signal_filename, background_filename):
    # Read data from ROOT files
    data_sig = ROOT.RDataFrame("Events", signal_filename).AsNumpy()
    data_bkg = ROOT.RDataFrame("Events", background_filename).AsNumpy()

    # Convert inputs to format readable by machine learning tools
    x_sig = np.vstack([data_sig[var] for var in variables]).T
    x_bkg = np.vstack([data_bkg[var] for var in variables]).T
    x = np.vstack([x_sig, x_bkg])

    # Create labels
    num_sig = x_sig.shape[0]
    num_bkg = x_bkg.shape[0]
    y = np.hstack([np.ones(num_sig), np.zeros(num_bkg)])

    # Compute weights balancing both classes
    num_all = num_sig + num_bkg
    w = np.hstack([np.ones(num_sig) * num_all / num_sig, np.ones(num_bkg) * num_all / num_bkg])

    return x, y, w

# Load data
x, y, w = load_data("train_signal.root", "train_background.root")

# Fit xgboost model
from xgboost import XGBClassifier
xbdt = XGBClassifier(max_depth=3, n_estimators=500)
xbdt.fit(x, y, sample_weight=w)

# Save model in TMVA format
ROOT.TMVA.Experimental.SaveXGBoost(xbdt, "myBDT", "tmva101.root", num_inputs=x.shape[1])

# Load data
x, y_true, w = load_data("test_signal.root", "test_background.root")

# Load trained model
File = "tmva101.root"

bdt = ROOT.TMVA.Experimental.RBDT("myBDT", File)

# Make prediction
y_pred_ref = xbdt.predict_proba(x)[:, 1]
y_pred = bdt.Compute(x).T

diff = y_pred - y_pred_ref

print("max diff", np.max(np.abs(diff)))
```
Before, the max difference was `0.379578`, and with this PR it is `5.9604645e-08`, compatible with floating point precision.

FYI @martamaja10, @lmoneta, @vepadulano

### CODE REVIEW

File: tmva/tmva/src/RBDT.cxx
Comment (hahnjo) [inline tmva/tmva/src/RBDT.cxx:9]: That doesn't look right...

File: tmva/tmva/src/RBDT.cxx
Comment (hahnjo) [inline tmva/tmva/src/RBDT.cxx:37]: You usually want `unordered_map`s because they have better runtime complexity, unless the code really needs the key ordering guarantees

File: tmva/tmva/src/RBDT.cxx
Comment (hahnjo) [inline tmva/tmva/src/RBDT.cxx:46]: This searches each map twice; you probably want to use `find()` and use the returned iterator directly

File: tmva/tmva/src/RBDT.cxx
Comment (hahnjo) [inline tmva/tmva/src/RBDT.cxx:42]: should be easier to write and read with a range-based `for` loop and `auto`

File: tmva/tmva/src/RBDT.cxx
Comment (hahnjo) [inline tmva/tmva/src/RBDT.cxx:188]: range-based `for` and `auto`

File: tmva/tmva/src/RBDT.cxx
Comment (hahnjo) [inline tmva/tmva/src/RBDT.cxx:282]: `std::unordered_map`

**Comment by hahnjo** [comment]
To be honest, I'm not convinced this should go into 6.32: After branching, feature development is nominally over...

**Comment by guitargeek** [comment]
But if the only tutorial for that feature silently gives wrong results? Personally I think that deserves an exception.

File: tmva/tmva/src/RBDT.cxx
Comment (guitargeek) [reply tmva/tmva/src/RBDT.cxx:9]: Thanks, nice catch!

File: tmva/tmva/src/RBDT.cxx
Comment (guitargeek) [reply tmva/tmva/src/RBDT.cxx:42]: Thanks! FYI, the original code doesn't use range-based loops and `auto` for C++98 compatibility

**Review by vepadulano — CHANGES_REQUESTED** [review]
Thank you! This will indeed be nice to have for the AGC implementation (although I am not sure yet if this is a good argument for bringing it to 6.32, the test silently returning wrong data is a much stronger argument).

There is some code modernization and class cleanup to be done (applying ROOT's camelcase everywhere for example).

File: tmva/tmva/inc/TMVA/RBDT.hxx
Comment (guitargeek) [reply tmva/tmva/inc/TMVA/RBDT.hxx:39]: It's also needed by the `txt` IO, so removing it fill make the configuration fail. I can add a comment about this though

**Comment by guitargeek** [comment]
Currently not:
```txt
/home/rembserj/code/root/tmva/tmva/src/RBDT.cxx:304:23: error: no matching function for call to ‘getline(const std::istream&, std::string&)’
  304 |    while (std::getline(file, line)) {
```

File: tmva/tmva/src/RBDT.cxx
Comment (guitargeek) [reply tmva/tmva/src/RBDT.cxx:275]: Currently not:
```txt
/home/rembserj/code/root/tmva/tmva/src/RBDT.cxx:304:23: error: no matching function for call to ‘getline(const std::istream&, std::string&)’
  304 |    while (std::getline(file, line)) {
```

File: tmva/tmva/src/RBDT.cxx
Comment (guitargeek) [reply tmva/tmva/src/RBDT.cxx:59]: I would like to keep the C-style interface for compatibility reasons with the original fastforest, in case I want to sync code again.

There is just no alternative to it yet. `std::array` and `std::vector` are too specific, and `std::span` only comes in C++20.

File: tmva/tmva/src/RBDT.cxx
Comment (guitargeek) [reply tmva/tmva/src/RBDT.cxx:205]: See this comment about array interface:
https://github.com/root-project/root/pull/15173#discussion_r1557418856

File: tmva/tmva/src/RBDT.cxx
Comment (guitargeek) [reply tmva/tmva/src/RBDT.cxx:218]: See this comment about array interface:
https://github.com/root-project/root/pull/15173#discussion_r1557418856

File: tmva/tmva/src/RBDT.cxx
Comment (guitargeek) [reply tmva/tmva/src/RBDT.cxx:149]: I'll use `TSystem` to be more consistent with the rest of ROOT. I think we had problems with `std::filesystem` on Windows.

**Comment by guitargeek** [comment]
Hi @vepadulano, thanks for your review comments! I have addressed all of them, except for the CamelCase in the implementation details. I think since this code is copied from an external library and I might want to sync it again at some point, it should be kept as-is for now. Would that be okay for you?

**Comment by vepadulano** [comment]
> Would that be okay for you?

I would still like the repository to use the same casing everywhere, but probably if we look hard enough we can find other examples of functions not strictly following the CamelCase convention.

For the purposes of this PR, as long as anything outside of `Internal` namespace or unnamed namespaces in compilation units follows CamelCase, I would be ok with it.

**Comment by guitargeek** [comment]
> For the purposes of this PR, as long as anything outside of `Internal` namespace or unnamed namespaces in compilation units follows CamelCase, I would be ok with it.

Yep that's the case! Everything not following the convention is in the anonymous namespace.

**Comment by guitargeek** [comment]
By the way, is xgboost installed in the new CI? If not, we are not testing these developments right now.

**Comment by vepadulano** [comment]
>  is xgboost installed in the new CI

If it's not in https://github.com/root-project/root/blob/master/requirements.txt then it's not installed

**Comment by guitargeek** [comment]
But should we explicitly add optional ROOT dependencies there? Not sure. Anyway I put it there temporarily, for the sake of a CI run.

**Comment by vepadulano** [comment]
> But should we explicitly add optional ROOT dependencies there?

That file contains only optional ROOT dependencies (i.e. runtime dependencies)

File: tmva/tmva/src/RBDT.cxx
Comment (vepadulano) [inline tmva/tmva/src/RBDT.cxx:41]: What is the meaning of the keys and values in this map type?

File: tmva/tmva/src/RBDT.cxx
Comment (vepadulano) [inline tmva/tmva/src/RBDT.cxx:72]: Resetting the value of `i` inbetween for loops means that probably it should just be declared in the for loop directly

File: tmva/tmva/test/CMakeLists.txt
Comment (vepadulano) [inline tmva/tmva/test/CMakeLists.txt:36]: You need to add the `rbdt_xgboost.py` file to the CMakeLists for it to be included. In the CMake command you should also state the dependency on `xgboost` so that ctest runs the corresponding fixture beforehand to check the package is present in the environment

**Review by vepadulano — CHANGES_REQUESTED** [review]

**Comment by hahnjo** [comment]
> But should we explicitly add optional ROOT dependencies there? Not sure. Anyway I put it there temporarily, for the sake of a CI run.

This won't work, dependencies from `requirements.txt` are installed directly into the CI image which is generated once per night based on `master`.

File: tmva/tmva/inc/TMVA/RBDT.hxx
Comment (guitargeek) [reply tmva/tmva/inc/TMVA/RBDT.hxx:75]: Since I originally followed the CMSSW code style, I used trailing underscores to mark public data members. Coincidentally, the xgboost python package follows almost the same convention (only with `snake_case_`, as one can see in examples like `xgb_model.n_classes_`).

The `fDataMember` convention is for private data members AFAIK. Anyway the best solution is probably to make all these data members private anyway, and then we can go for `fStyle`.

**Comment by guitargeek** [comment]
Ok thanks! Note that I'll also have to change the order of imports in the tutorials because of this issue, which was independently reproduced by @vepadulano :
  * https://github.com/wlav/cppyy/issues/227

**Comment by martamaja10** [comment]
With help from Jonas, I tested this PR in the context of AGC (for now locally on my PC with 8 cores) and it works - the ml histograms are produced.

File: tmva/tmva/src/RBDT.cxx
Comment (guitargeek) [reply tmva/tmva/src/RBDT.cxx:41]: I have added some docs now, but I think it's not necessary to go into detail here.

In the future, I think we can get rid of it anyway: at some point the XGBoost devs took inspiration from my library to optimize their own BDT representation, if I remember the discussions on GitHub correctly.

So I think we can avoid this index translation in the future by parsing XGBoosts internal representation directly instead of going over the `txt` dump.

**Comment by guitargeek** [comment]
Thanks Jonas for the valid concern.

There are some people who tried to use RBDT, and even people who noticed the problem or wrong values for XGBoost models:

https://root-forum.cern.ch/search?q=Rbdt

Also, the plan to "consolidate RBDT" was announced in the PoW and at several presentations, so at this point dropping the feature is not an option I think.

I agree with you that if a feature gave wrong results in the tutorial and nobody opened an issue about it for 5 years: doesn't exactly speak for the relevance of the feature.

**Comment by guitargeek** [comment]
Actually I think users don't use the GitHub issue tracker that much in general. It's mostly used by devs, packagers, and experiment framework developers. So maybe it's normal we didn't hear about the problem there.

**Comment by vepadulano** [comment]
I have summarized the current understanding of the situation with RBDT at https://github.com/root-project/root/issues/15197

Given that, I believe we should not merge this at the moment (and thus not push for it being in 6.32). An investigation is required regarding how to provide our Python users with what they probably want in this case: evaluating xgboost models from input data coming from a ROOT file (e.g. TTree or RNTuple). This may be served by the changes of this PR. But an alternative is also forgoing an internal implementation a-la RBDT and just try to see if we can hook the xgboost Python API directly into e.g. the RDataFrame API.

**Comment by guitargeek** [comment]
Awesome, thanks!

> But an alternative is also forgoing an internal implementation a-la RBDT and just try to see if we can hook the xgboost Python API directly into e.g. the RDataFrame API.

I don't think that the XGBoost API is thread safe (it doesn't have to be because of the GIL), so that might be a problem with implicit multi-threading in RDataFrame.

I'm happy to continue the discussion later, but just for the record: I would still consider this a bugfix for 6.32.

**Comment by guitargeek** [comment]
To summarize what was decided today in the ROOT team meeting: it is very unfortunate that this bug was discovered a week after branching, please avoid this in the future. Nevertheless, since this is fixing a feature that we advertise and the feature is very specific, this should be merged and backported to 6.32 if the CI can be made green (even disabling the failing test would be acceptable, since n - 1 tested platforms are better than zero).

**Comment by guitargeek** [comment]
I have now installed the final review comments about the `fMember` style by @vepadulano.

Furthermore, the `alma9` failures should be gone now, because we have deactivated `tmva-cpu` for this platform. Like this, the collision of the openblas version on the system (picked up by TMVA-CPU) and by NumPy should not happen.

File: tmva/tmva/inc/TMVA/RBDT.hxx
Comment (guitargeek) [reply tmva/tmva/inc/TMVA/RBDT.hxx]: I have made all the data members private and use the `fDataMember` style in my latest update!

**Review by lmoneta — APPROVED** [review]
LGTM!
Thanks a lot Jonas for this nice contribution

**Pull request merged**
